src/modules/dwm.cpp

```diff
@@ -0,0 +1,493 @@
+        m_ipc->subscribe(dwmipc::Event::TAG_CHANGE);
+        m_log.info("%s: Initialized tags", name());
+      }
+
+      if (m_layout_label) {
+        m_layouts = m_ipc->get_layouts();
+
+        // First layout is treated as default by dwm
+        m_default_layout = &m_layouts->at(0);
+        m_current_layout = find_layout(m_bar_mon->layout.address.cur);
+
+        if (m_layout_click) {
+          m_secondary_layout = find_layout(m_secondary_layout_symbol);
+
+          if (m_secondary_layout == nullptr) {
+            throw module_error("Secondary layout symbol does not exist");
+          }
+        }
+
+        // Initialize layout symbol
+        update_layout_label(m_bar_mon->layout.symbol.cur);
+
+        // This event is only needed to update the layout label
+        m_ipc->on_layout_change = [this](const dwmipc::LayoutChangeEvent& ev) { on_layout_change(ev); };
+        m_ipc->subscribe(dwmipc::Event::LAYOUT_CHANGE);
+        m_log.info("%s: Initialized layout label", name());
+      }
+
+      // These events are only necessary to update the focused window title
+      if (m_title_label) {
+        update_title_label();
+        m_ipc->on_client_focus_change = [this](const dwmipc::ClientFocusChangeEvent& ev) {
+          this->on_client_focus_change(ev);
+        };
+        m_ipc->subscribe(dwmipc::Event::CLIENT_FOCUS_CHANGE);
+
+        m_ipc->on_focused_title_change = [this](const dwmipc::FocusedTitleChangeEvent& ev) {
+          this->on_focused_title_change(ev);
+        };
+        m_ipc->subscribe(dwmipc::Event::FOCUSED_TITLE_CHANGE);
+        m_log.info("%s: Initialized title label", name());
+      }
+
+      if (m_floating_label) {
+        update_floating_label();
+        m_ipc->on_focused_state_change = [this](const dwmipc::FocusedStateChangeEvent& ev) {
+          this->on_focused_state_change(ev);
+        };
+        m_ipc->subscribe(dwmipc::Event::FOCUSED_STATE_CHANGE);
+        m_log.info("%s: Initialized floating label", name());
+      }
+
+      // This event is for keeping track of the currently focused monitor
+      m_ipc->on_monitor_focus_change = [this](const dwmipc::MonitorFocusChangeEvent& ev) {
+        this->on_monitor_focus_change(ev);
+      };
+      m_ipc->subscribe(dwmipc::Event::MONITOR_FOCUS_CHANGE);
+        m_log.info("%s: Subscribed to monitor focus change", name());
+    } catch (const dwmipc::IPCError& err) {
+      throw module_error(err.what());
+    }
+  }
+
+  void dwm_module::stop() {
+    try {
+      m_log.info("%s: Disconnecting from socket", name());
+      m_ipc.reset(nullptr);
+    } catch (...) {
+    }
+
+    event_module::stop();
+  }
+
+  auto dwm_module::has_event() -> bool {
+    try {
+      return m_ipc->handle_event();
+    } catch (const dwmipc::SocketClosedError& err) {
+      m_log.err("%s: Disconnected from socket: %s", name(), err.what());
+      sleep(chrono::duration<double>(1));
```

Comment:
I don't think we should sleep here because we have a lock on `m_updatelock`. At some point we will combine `m_updatelock` and `m_buildlock` into a single mutex to fix some concurrency issues and then sleeping here may block the entire bar.

We could sleep in the `idle` method and schedule a reconnect in the next call to `has_event`. I'm not sure we can reconnect in `idle` because we may run into some concurrency issues.

Reply:
Ok, I'll look into refactoring the disconnects and look more into how the different threads operate.

Reply:
Let me know if you are stuck anywhere. The concurrency model of polybar isn't exactly transparent, I get confused quite often as well.
I think in the future, we should just put a big lock on anything that calls into the module and on the main module loop as well.

Reply:
See commits 4a5070ce6c8a152be0945afb7d6bde2ed269315f and b1caed225b660db2edc420e03e504af5650d660c